### DISCUSSION TITLE
Blank or empty window when building it from source in arch Linux

### DISCUSSION THREAD

**Discussion by topas-rec** [pulsar-edit/.github#262] [Core Application & Packages]
Hello everyone,

I an currently creating an AUR package that builds from source for arch Linux.

It get it to build and launch, but it is... empty, when launched without presenting much information about why.

![Screenshot_20240322-085547_aRDP Free](https://github.com/pulsar-edit/.github/assets/20391449/5af05fe2-cd1c-4f0b-900d-833bd55e7aa3)

It says `(node:87717) Electron: Loading non-context-aware native module in renderer: '/tmp/.org.chromium.Chromium.jLeX3s'. This is deprecated, see https://github.com/electron/electron/issues/18397.` on the command line when started, but I don't think that had something to do with the problem.

I tried running it from the binaries created for Debian, from the deb archive, I tried to run the app (app.asar) with a system electron (version 12) with the same results.
[arch wiki - electron packaging guidelines](https://wiki.archlinux.org/title/Electron_package_guidelines#Using_the_system_electron)

**Comment by confused-Techie** [comment]
Did you try running with `--no-sandbox` or `--in-process-gpu`?

This is a known issue on some Linux distributions, and is generally included within the released binary of them.

**Reply by topas-rec** [reply]
Just tried that and these options don't help.

I was able to open the devtools and I think they deliver something:

![Screenshot from 2024-03-22 20-51-33](https://github.com/pulsar-edit/.github/assets/20391449/c560758b-d94d-4fe7-af75-0a0ce64bb38a)

```
electron/js2c/renderer_init.js:13 
(electron) The remote module is deprecated. Use https://github.com/electron/remote instead.
log @ electron/js2c/renderer_init.js:13

electron/js2c/renderer_init.js:113
Electron Security Warning (Insecure Content-Security-Policy) This renderer process has either no Content Security
    Policy set or a policy with "unsafe-eval" enabled. This exposes users of
    this app to unnecessary security risks.

For more information and help, consult
https://electronjs.org/docs/tutorial/security.
This warning will not show up
once the app is packaged.
(anonymous) @ electron/js2c/renderer_init.js:113
electron/js2c/renderer_init.js:33

@ electron/js2c/renderer_init.js:33
The vm module of Node.js is deprecated in the renderer process and will be removed.
o._load

index.js:83
Error: /var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/node_modules/git-utils/build/Release/git.node: undefined symbol: git_net_url_is_default_port
    at process.func [as dlopen] (electron/js2c/asar_bundle.js:5)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1138)
    at Object.func [as .node] (electron/js2c/asar_bundle.js:5)
    at Module.load (internal/modules/cjs/loader.js:935)
    at Module._load (internal/modules/cjs/loader.js:776)
    at Function.f._load (electron/js2c/asar_bundle.js:5)
    at Function.o._load (electron/js2c/renderer_init.js:33)
    at Module.require (internal/modules/cjs/loader.js:959)
    at require (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/native-compile-cache.js:67)
    at Object.<anonymous> (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/node_modules/git-utils/src/git.js:3)
    at Object.<anonymous> (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/node_modules/git-utils/src/git.js:358)
    at Module._compile (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/native-compile-cache.js:120)
    at Object.value [as .js] (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/compile-cache.js:252)
    at Module.load (internal/modules/cjs/loader.js:935)
    at Module._load (internal/modules/cjs/loader.js:776)
    at Function.f._load (electron/js2c/asar_bundle.js:5)
    at Function.o._load (electron/js2c/renderer_init.js:33)
    at Module.require (internal/modules/cjs/loader.js:959)
    at require (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/native-compile-cache.js:67)
    at Object.<anonymous> (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/git-repository.js:5)
    at Object.<anonymous> (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/git-repository.js:623)
    at Module._compile (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/native-compile-cache.js:120)
    at Object.value [as .js] (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/compile-cache.js:252)
    at Module.load (internal/modules/cjs/loader.js:935)
    at Module._load (internal/modules/cjs/loader.js:776)
    at Function.f._load (electron/js2c/asar_bundle.js:5)
    at Function.o._load (electron/js2c/renderer_init.js:33)
    at Module.require (internal/modules/cjs/loader.js:959)
    at require (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/native-compile-cache.js:67)
    at Object.<anonymous> (/var/lib/archbuild/extra-x86_64/tobiasb/build/pulsar-editor/src/pulsar/src/git-repository-provider.js:3)
handleSetupError @ index.js:83
```

I also checked that running with `yarn start` it also doesn't work.

**Reply by confused-Techie** [reply]
That does help quite a lot.
It looks like for some reason your native modules are failing to build properly.

I'd double check that the versions of other applications are the same as what we use in our build files. I'd recommend looking at [`build.yml`](https://github.com/pulsar-edit/pulsar/blob/master/.github/workflows/build.yml) GHA, and the [`cirrus.yml`](https://github.com/pulsar-edit/pulsar/blob/master/.cirrus.yml) Cirrus CI config. Cirrus might provide more help since those images don't include much of anything so we have to spell out every install manually, meanwhile the GHA one of course uses GHA images which are full of applications by default.

Otherwise maybe it's helpful to look at how an [existing AUR](https://aur.archlinux.org/packages/pulsar-bin) build was done?

Sorry I can't offer more advice, I'm not to familiar with Archi Linux nor the native module building process beyond a very superficial amount of knowledge. But hopefully those resources help

**Reply by topas-rec** [reply]
Great.

I don't alter the files you mention, just take the release source code from https://github.com/pulsar-edit/pulsar/releases/tag/v1.114.0 and build it using the descriptions in the pulsar docs.
When I do not alter the files, does it make sense to look into them?

The existing AUR doesn't build, it takes the binary release and installed them on arch (that's why it is named `-bin`.

Here is the PKGBUILD which contains the build commands.
The build function is the important part.

```
# Maintainer: Tobias Boesch
# https://wiki.archlinux.org/title/Creating_packages#More_detailed_guidelines
#   https://wiki.archlinux.org/title/Electron_package_guidelines
pkgname=pulsar-editor
pkgver=v1.114.0
_pkgver_deb_file=${pkgver:1} # https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameter-Expansion
_repo_name=${pkgname:0:6}
pkgrel=1
pkgdesc="A Community-led Hyper-Hackable Text Editor"
arch=('x86_64')
url="https://pulsar-edit.dev"
_repo_url=github.com/pulsar-edit
license=('MIT')
groups=()
depends=('libxext'
          'libdrm'
          'libxcomposite'
          'glib2'
          'libxkbcommon'
          'gcc-libs'
          'gdk-pixbuf2'
          'python'
          'libxfixes'
          'nss'
          'curl'
          'cairo'
          'libxshmfence'
          'libx11'
          'nspr'
          'libxrandr'
          'python-setuptools'
          'gtk3'
          'libsecret'
          'glibc'
          'libcups'
          'libxshmfence'
          'libxcomposite'
          'at-spi2-core'
          'gdk-pixbuf2'
          'mesa'
          'libxcb'
          'expat'
          'bash'
          'libxdamage'
          'pango'
          'alsa-lib'
          'perl')
# https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#requirements-and-dependencies
makedepends=('libxkbfile'
              'libsecret'
              'libx11'
              'libxcrypt-compat'
              'git'
              'nodejs'
              'yarn'
              'python'
              'gcc12'
              'nvm')
optdepends=()
provides=()
conflicts=(pulsar-bin)
replaces=()
backup=()
options=()
install=
changelog=
source=(git+https://$_repo_url/$_repo_name.git#tag=$pkgver)
noextract=()
sha256sums=(SKIP) #autofill using updpkgsums

# https://wiki.archlinux.org/title/Node.js_package_guidelines#Using_nvm
_ensure_local_nvm() {
    # let's be sure we are starting clean
    which nvm >/dev/null 2>&1 && nvm deactivate && nvm unload
    export NVM_DIR="${srcdir}/.nvm"

    # The init script returns 3 if version specified
    # in ./.nvrc is not (yet) installed in $NVM_DIR
    # but nvm itself still gets loaded ok
    source /usr/share/nvm/init-nvm.sh || [[ $? != 1 ]]
}

build() {
  # https://github.com/pulsar-edit/pulsar/issues/684#issuecomment-1749267403
  export CC=gcc-12 CXX=g++-12

  cd "$_repo_name"
  _ensure_local_nvm

  # https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#building-and-running-the-application
  nvm install
  corepack enable

  nvm use
  node -v

  git submodule init && git submodule update

  yarn install

  yarn build
  yarn build:apm

  # https://pulsar-edit.dev/docs/launch-manual/sections/core-hacking/#building-binaries
  yarn dist deb
}

package() {

  # https://duckduckgo.com/?q=yarn+site%3Aaur.archlinux.org%2Fcgit%2F&t=ffab&ia=web
  #   https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=jellyseerr
  mv ${_repo_name}/dist/*.deb "${srcdir}/"
  cd "${srcdir}"
  ar x ${_repo_name}_${_pkgver_deb_file}_amd64.deb data.tar.xz
  bsdtar -xf ${srcdir}/data.tar.xz -C ${pkgdir}/

  # https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=pulsar-bin
  
  # mv opt "$pkgdir"
  # mv usr "$pkgdir"

  # # Cleanup specs. Remove if implemented upstream
  # find "$pkgdir/opt/Pulsar/resources/app/ppm" -type d -name "spec" -exec rm -rf {} +
  # find "$pkgdir/opt/Pulsar/resources/app.asar.unpacked" -type d -name "spec" -exec rm -rf {} +

  # install -Dm644 "$pkgdir/opt/Pulsar/resources/pulsar.svg" -t "$pkgdir/usr/share/icons/hicolor/scalable/apps"

  # install -Dm755 -d "$pkgdir/usr/bin"
  # chmod +x "$pkgdir/opt/Pulsar/resources/pulsar.sh"

  # ln -sf "/opt/Pulsar/resources/pulsar.sh" "$pkgdir/usr/bin/pulsar"
}

```

**Reply by DeeDeeG** [reply]
Hi, if you look a little closer at the above `build()` function, it is using our build system to create a packaged `.deb`, then it extracts (`ar`) the things from the `.deb` that it needs from it.

This is building from source.

**Reply by DeeDeeG** [reply]
Hmm, sorry, this is _your own work_, isn't it? My apologies.

**Comment by topas-rec** [comment]
> Otherwise maybe it's helpful to look at how an [existing AUR](https://aur.archlinux.org/packages/pulsar-bin) build was done?

But yeah, there are other builds that are non binary - I'll check out how they do it.
I already saw they set the "don't strip" option. Maybe that's the reason.

**Comment by topas-rec** [comment]
The striping of symbols wasn't the problem.
How can I debug the missing symbol? Where should it be build and how can I find out why it is not build correctly?

**Comment by topas-rec** [comment]
> When I do not alter the files, does it make sense to look into them?

Saw that those are pipeline build files. Does make sense.

I looked into your pipeline logs and I see that I am missing a warning that you have in your logs:

```
In file included from /github/home/.electron-gyp/12.2.3/include/node/node.h:67,
                 from ../../../nan/nan.h:60,
                 from ../src/binding.cpp:1:
/github/home/.electron-gyp/12.2.3/include/node/v8.h:1650:79: warning: 'ResolveCallback' is deprecated [-Wdeprecated-declarations]
                                                       ResolveCallback callback);
                                                  
```

Maybe I use a newer `node-gyp` version.
I tried to install it using `npx --yes node-gyp install 16` but it doesn't change a thing.

I am attaching the build log. Perhaps you find something wrong in it.
[pulsar-editor-v1.115.0-1-x86_64-build.log](https://github.com/pulsar-edit/.github/files/14771487/pulsar-editor-v1.115.0-1-x86_64-build.log)

**Reply by confused-Techie** [reply]
Sorry for not having the time to respond earlier.

But the reason I pointed you to those files was to recommend that you follow them as closely as you possible can. The versions of many of our dependencies are exact and without them or even slightly different versions will result in Pulsar building incorrectly.

Especially take special care for the following:
* Version of NodeJS
* Version of `node-gyp`
* Version of Python
* Version of NPM
* Ensuring to build PPM 
* Ensuring to redownload proper v8 header files

But I did look into the build log, and without diving too deep into it seems that everything is looking pretty good. As far as I can tell you've gotten the versioning on everything correct, and the build for the most part seems identical, besides some output on the Native Module builds, but nothing that'd make me think it'd start to crash everything. 

Admittedly Linux isn't my strong suit when it gets to the nitty gritty like this, I'd be curious if anyone over at @pulsar-edit/core has any further ideas on what could be going wrong here.

**Comment by ghost** [comment]
did you try building it separately and zipping with makepkg
- Build from source without makepkg (Directly in repo yarn dist **)
- zip it with makepkg (similar to  [pulsar-bin](https://aur.archlinux.org/packages/pulsar-bin))
**My Repo** https://github.com/Ivy-Tokito/pulsar-arch-package.git
it's working for me
